esi_leap/db/sqlalchemy/api.py

```diff
@@ -133,23 +133,139 @@ def offer_get_all(filters):
     start = filters.pop('start_time', None)
     end = filters.pop('end_time', None)
 
+    a_start = filters.pop('available_start_time', None)
+    a_end = filters.pop('available_end_time', None)
+
+    if a_start and a_end and a_end <= a_start:
+        raise exception.InvalidAvailabilityCommand(a_start=a_start,
+                                                   a_end=a_end)
+
     query = query.filter_by(**filters)
 
     if start and end:
         query = query.filter((start >= models.Offer.start_time) &
                              (end <= models.Offer.end_time))
 
+    for o in query:
+        try:
+            offer_verify_availability(o, start_time=a_start, end_time=a_end)
+        except exception.OfferNotAvailable:
+            query = query.filter(models.Offer.uuid != o.uuid)
+
     return query
 
 
+def offer_get_availabilities(offer_ref):
+
+    c_query = model_query(models.Contract, get_session())
+    availabilities_ref = c_query.with_entities(models.Contract.start_time, models.Contract.end_time).\
+        join(models.Offer).\
+        order_by(models.Contract.start_time).\
+        filter(models.Contract.offer_uuid == offer_ref.uuid).all()
+
+    if availabilities_ref:
+        a = [offer_ref.start_time, availabilities_ref[0][0]]
+        for i in range(len(availabilities_ref) - 1):
+            a.append(availabilities_ref[i][1])
+            a.append(availabilities_ref[i + 1][0])
+        a.append(availabilities_ref[-1][1])
+        a.append(offer_ref.end_time)
+
+        i = 0
+        while i < len(a) - 1:
+            if a[i] == a[i + 1]:
+                a.pop(i)
+                a.pop(i)
+            else:
+                i += 1
+
+        a = [[a[j], a[j + 1]] for j in range(0, len(a) - 1, 2)]
+
+    else:
+        a = [[offer_ref.start_time, offer_ref.end_time]]
+
+    return a
+
+
+def offer_verify_availability(offer_ref, start_time=None, end_time=None):
```

Comment:
This feels like it can be done through a single filtered query on contracts - something like:
```
select * from contracts c where c.offer_id = _offer_id_ and ((c.start_time <= _start_time_ and c.end_time >= _start_time_) or (c.end_time >= _end_time_ and c.start_time <= _end_time_))
```

Reply:
This cannot! The problem is the start_times and end_times in the db are not availability start and end_times, but rather UN-availability start and end times. I tried to see if I could do this in the db but it seemed impossible

Reply:
Right - but finding a conflicting contract - like through the query above - and saying "if one exists, then this isn't available" - feel like it should work? Or am I missing something?

Also, start_time/end_time should always be specified for this function, right?

Reply:
This method does two things. It verifies if a start and end time are valid, and it finds the start and end time for contracts that did not pass in both parameters explicitly. Originally I had it so if you passed in a start time but no end time for contract creation, it grabbed the first availabilities end time. But I realized it should instead return the end time of the availability slot in which start_time lives.

Reply:
The checking validity could be done in the db, but the returning times part could not I think.

Reply:
Okay - those two functions should be split. At first I thought that maybe this entire function should be moved out of the db layer; then I realized you were treating the start_time/end_time validity check as a database integrity issue, which is definitely fair. So that part should stay here, and probably be simplified into the kind of database query I expressed above.

The other half - finding start/end times if unspecified - should be separated out, and allowed to function in the object layer. Honestly, I'm starting to lean towards cutting this functionality from this patch for being overly complex. At most, I'd suggest the following:

* if start_time is unspecified, set it to current_time in the REST API layer
* if end_time is unspecified, look for the earliest start_time > current_time from a related contract (this could be grabbing the first contract from a filtered contract query); set this at the REST API layer
* let your database start/end time validation deal with the rest

Reply:
I mean the work for no start or end time is there. Why not keep it? I can move the validity checks to other functions in the db however.

Reply:
If it's not doing anything in the current code base, we don't want it in there as extra clutter. I'd keep it in a branch for now.